lib/copy/copy.js

```diff
@@ -2,51 +2,211 @@
+
+  // Warn about using preserveTimestamps on 32-bit node
   if (options.preserveTimestamps && process.arch === 'ia32') {
     console.warn(`fs-extra: Using the preserveTimestamps option in 32-bit node is not recommended;\n
-    see https://github.com/jprichardson/node-fs-extra/issues/269`)
+      see https://github.com/jprichardson/node-fs-extra/issues/269`)
   }
 
   // don't allow src and dest to be the same
   const basePath = process.cwd()
-  const currentPath = path.resolve(basePath, src)
-  const targetPath = path.resolve(basePath, dest)
-  if (currentPath === targetPath) return callback(new Error('Source and destination must not be the same.'))
-
-  fs.lstat(src, (err, stats) => {
-    if (err) return callback(err)
-
-    let dir = null
-    if (stats.isDirectory()) {
-      const parts = dest.split(path.sep)
-      parts.pop()
-      dir = parts.join(path.sep)
-    } else {
-      dir = path.dirname(dest)
+  src = path.resolve(basePath, src)
+  dest = path.resolve(basePath, dest)
+  if (src === dest) return cb(new Error('Source and destination must not be the same.'))
+
+  if (options.filter) {
+    if (options.filter instanceof RegExp) {
+      console.warn('Warning: fs-extra: Passing a RegExp filter is deprecated, use a function')
+      if (!options.filter.test(src)) return cb()
+    } else if (typeof options.filter === 'function') {
+      if (!options.filter(src, dest)) return cb()
+    }
+  }
+
+  const destParent = path.dirname(dest)
+  mkdirs(destParent, err => {
+    if (err) return cb(err)
+
+    let stat = options.dereference ? fs.stat : fs.lstat
+    stat(src, (err, srcStat) => {
+      if (err) return cb(err)
+
+      if (srcStat.isDirectory()) {
+        return onDir(src, srcStat, dest, options, cb)
+      } else if (srcStat.isFile() || srcStat.isCharacterDevice() || srcStat.isBlockDevice()) {
+        return onFile(src, srcStat, dest, options, cb)
+      } else if (srcStat.isSymbolicLink()) {
+        return onLink(src, dest, options, cb)
+      }
+    })
+  })
+}
+
+function checkDest (dest, cb) {
+  fs.readlink(dest, (err, resolvedDestPath) => {
+    if (err) {
+      if (err.code === 'ENOENT') return cb(null, DEST_NOENT)
+      // dest exists but is not a link
+      if (err.code === 'EINVAL' || err.code === 'UNKNOWN') return cb(null, DEST_EXISTS)
+      return cb(err)
+    }
+    // dest exists and is a link
+    return cb(null, resolvedDestPath)
+  })
+}
+
+function onFile (src, srcStat, dest, options, cb) {
+  checkDest(dest, (err, res) => {
+    if (err) return cb(err)
+    if (res === DEST_NOENT) {
+      return cpFile(src, srcStat, dest, options, cb)
+    } else if (res === DEST_EXISTS) {
+      if (options.overwrite) {
+        fs.unlink(dest, err => {
+          if (err) return cb(err)
+          return cpFile(src, srcStat, dest, options, cb)
+        })
+      } else if (options.errorOnExist) {
+        return cb(new Error(dest + ' already exists'))
+      } else return cb()
+    } else return cb()
```

Comment:
This last `else` basically means dest exists and is a symlink. I believe the same pattern that is used for the case `DEST_EXISTS`, which means checking for `options.overwrite` and operates accordingly, should be applied for this case as well. Is that right?